apps/aecore/src/aec_fork_signalling.erl

```diff
@@ -0,0 +1,350 @@
+        ]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2
+        ]).
+
+-export([worker_process/4]).
+
+-define(SERVER, ?MODULE).
+
+-type header_hash()      :: aec_blocks:block_header_hash().
+
+-type worker()           :: {pid(), reference()}.
+
+-type result()           :: true | false | pending.
+
+%% The pointer is a key block hash which belongs to the last signalling block.
+-type pointer()          :: header_hash().
+
+-record(state, {fork     :: aec_hard_forks:fork() | undefined,
+                results  :: #{header_hash() := result() | pointer()},
+                workers  :: #{worker() := header_hash()}
+               }).
+
+-define(IS_RESULT(X), X =:= true orelse X =:= false orelse X =:= pending).
+-define(IS_POINTER(X), is_binary(X)).
+
+%% API
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+get_fork_result(Block, BlockHash) ->
+    gen_server:call(?SERVER, {get_fork_result, Block, BlockHash}).
+
+compute_fork_result(Block, BlockHash) ->
+    gen_server:cast(?SERVER, {compute_fork_result, Block, BlockHash}).
+
+%% gen_server callbacks
+
+init([]) ->
+    process_flag(trap_exit, true),
+    {ok, #state{fork = aec_hard_forks:fork(), results = maps:new(),
+                workers = maps:new()}}.
+
+handle_call({get_fork_result, Block, BlockHash}, _From, State) ->
+    {Reply, State1} = handle_get_fork_result(Block, BlockHash, State),
+    {reply, Reply, State1};
+handle_call(_Request, _From, State) ->
+    {reply, ingnored, State}.
+
+handle_cast({compute_fork_result, Block, BlockHash}, State) ->
+    {noreply, handle_compute_fork_result(Block, BlockHash, State)};
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+handle_info({worker_msg, Pid, Msg}, State) ->
+    {noreply, handle_worker_msg(Pid, Msg, State)};
+handle_info({'DOWN', Ref, process, Pid, Rsn}, State) ->
+    {noreply, handle_down({Pid, Ref}, Rsn, State)};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Rsn, #state{workers = Workers} = State) ->
+    maps:foldl(fun(W, _BH, S1) -> kill_worker(W, S1) end, State, Workers).
+
+%% Internal functions
+
+handle_get_fork_result(Block, BlockHash,
+                       #state{fork = Fork, results = Results} = State) ->
+    case is_fork_height_block(Block, Fork) of
+        true ->
+            case get_result(BlockHash, Results) of
+                P when ?IS_POINTER(P) ->
+                    %% The block at height H2 is supposed to have a pointer
+                    %% to the block hash at height HE - 1 with the actual
+                    %% result.
+                    case get_result(P, Results) of
+                        R when ?IS_RESULT(R) ->
+                            {{ok, R}, State};
+                        _Other ->
+                            {{error, not_found}, State}
+                    end;
+                _Other ->
```

Comment:
If node is stopped when top is H2-1, when node starts which API of this module is meant to be used to make sure that computation of outcome of signalling is started?

Reply:
Good catch, will fix.

Reply:
Fixed in 4d496b383